app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -38,287 +35,134 @@ async function addEthereumChainHandler(
   end,
   {
     upsertNetworkConfiguration,
-    getCurrentChainId,
     getCurrentRpcUrl,
     findNetworkConfigurationBy,
-    setNetworkClientIdForDomain,
     setActiveNetwork,
     requestUserApproval,
     startApprovalFlow,
     endApprovalFlow,
-    getProviderConfig,
-    hasPermissions,
+    getCurrentChainIdForDomain,
+    getCaveat,
+    requestPermittedChainsPermission,
+    getChainPermissionsFeatureFlag,
   },
 ) {
-  if (!req.params?.[0] || typeof req.params[0] !== 'object') {
-    return end(
-      ethErrors.rpc.invalidParams({
-        message: `Expected single, object parameter. Received:\n${JSON.stringify(
-          req.params,
-        )}`,
-      }),
-    );
+  let validParams;
+  try {
+    validParams = validateAddEthereumChainParams(req.params[0], end);
+  } catch (error) {
+    return end(error);
   }
 
-  const { origin } = req;
-
   const {
     chainId,
-    chainName = null,
-    blockExplorerUrls = null,
-    nativeCurrency = null,
-    rpcUrls,
-  } = req.params[0];
+    chainName,
+    firstValidBlockExplorerUrl,
+    firstValidRPCUrl,
+    ticker,
+  } = validParams;
+  const { origin } = req;
 
-  const otherKeys = Object.keys(
-    omit(req.params[0], [
-      'chainId',
-      'chainName',
-      'blockExplorerUrls',
-      'iconUrls',
-      'rpcUrls',
-      'nativeCurrency',
-    ]),
+  const currentChainIdForDomain = getCurrentChainIdForDomain(origin);
+  const currentNetworkConfiguration = findExistingNetwork(
```

Comment:
Heads up that this will change following https://github.com/MetaMask/core/pull/4286, and you'll probably want something like:

``` typescript
const currentNetworkConfiguration = getNetworkConfigurationByChainId(chainIdForDomain);
```

(and similar for below)